ngraph/core/src/op/roll.cpp

```diff
@@ -0,0 +1,79 @@
+//*****************************************************************************
+// Copyright 2017-2021 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#include <ngraph/validation_util.hpp>
+
+#include "itt.hpp"
+#include "ngraph/op/roll.hpp"
+
+using namespace std;
+using namespace ngraph;
+
+NGRAPH_RTTI_DEFINITION(op::v7::Roll, "Roll", 7);
+
+op::v7::Roll::Roll(const Output<Node>& data, const Output<Node>& shift, const Output<Node>& axes)
+    : Op({data, shift, axes})
+{
+    constructor_validate_and_infer_types();
+}
+
+void op::v7::Roll::validate_and_infer_types()
+{
+    NGRAPH_OP_SCOPE(v7_Roll_validate_and_infer_types);
+
+    const auto& shift_et = get_input_element_type(1);
+    NODE_VALIDATION_CHECK(this,
+                          shift_et.is_dynamic() || shift_et.is_integral_number(),
+                          "Shift must have an integral number element type.");
+
+    const auto& axes_et = get_input_element_type(2);
+    NODE_VALIDATION_CHECK(this,
+                          axes_et.is_dynamic() || axes_et.is_integral_number(),
+                          "Axes must have an integral number element type.");
+
+    const auto& shift_pshape = get_input_partial_shape(1);
+    const auto& axes_pshape = get_input_partial_shape(2);
+    const auto& shift_rank = shift_pshape.rank().get_length();
+    const auto& axes_rank = axes_pshape.rank().get_length();
+
+    NODE_VALIDATION_CHECK(this, shift_rank <= 1, "Shift must be a scalar or 1D tensor.");
+
+    NODE_VALIDATION_CHECK(this, axes_rank <= 1, "Axes must be a scalar or 1D tensor.");
+
+    if (!(shift_pshape.is_static() &&
```

Comment:
This check looks too complicated to me. I understand that it goes for https://github.com/openvinotoolkit/openvino/pull/4469/files#diff-2b0119dc5b43a7eb3a925bfddabd9f0fb59890203d7fa0fbe1b0375c706bb3f5R58

AFAIU this if statement means:
>No need to check `shift` shape consistency with `axes` if `shift` is a scalar or 1D tensor with a single value . And always check consistency if `shift` partial_shape is dynamic.

Maybe adding comment will be appropriate but i'm not insisting on that. 

What is more important is that in the code scalar `shift` and `shift` with shape [1] behave the same way but spec says:
>If `shift` is a 1D tensor, `axes` must be a 1D tensor of the same size

If I understood correctly either the code should be corrected or the specification.

Reply:
Added comment and removed 1d tensor with a single value check.